extension/background.js

```diff
@@ -0,0 +1,230 @@
+    const patterns = Object.keys(matchPatternsMap);
+
+    if (patterns.length === 0) {
+      console.debug(
+        "not registering any listener because there is no URL to monitor"
+      );
+      return;
+    }
+
+    console.debug("registering onBeforeRequest listener");
+    browser.webRequest.onBeforeRequest.addListener(
+      this.webRequestHandler,
+      { urls: patterns },
+      ["blocking"]
+    );
+
+    // This one is needed in addition to `onBeforeRequest` because this
+    // extension might be registered before or after some extensions.
+    // Depending on that, our `onBeforeRequest` listener might not be called,
+    // which is why we also listen to `onBeforeRedirect`.
+    console.debug("registering onBeforeRedirect listener");
+    browser.webRequest.onBeforeRedirect.addListener(this.webRequestHandler, {
+      urls: patterns,
+    });
+  }
+
+  // `redirectUrl` is usually valid when the redirect has been detected via
+  // `onBeforeRedirect`, otherwise it's likely `undefined`.
+  webRequestHandler = async ({ requestId, url, redirectUrl }) => {
+    if (this.lastRequestIdReported === requestId) {
+      console.debug(`request ID '${requestId}' already reported, skipping...`);
+      return;
+    }
+
+    // When we detect a redirect, we read the request property, hoping to find
+    // an add-on ID corresponding to the add-on that initiated the redirect.
+    // It might not return anything when the redirect is a search server-side
+    // redirect but it can also be caused by an error.
+    let addonId = await browser.addonsSearchExperiment.getRequestProperty(
+      requestId,
+      "redirectedByExtension"
+    );
+
+    // When we did not find an add-on ID in the request property bag and the
+    // `redirectUrl` is both valid and different than the original URL. we
+    // likely detected a search server-side redirect.
+    const isServerSideRedirect =
+      !addonId && typeof redirectUrl !== "undefined" && url !== redirectUrl;
+
+    // Search server-side redirects are possible because an extension has
+    // registered a search engine, which is why we can (hopefully) retrieve the
+    // add-on ID.
+    if (isServerSideRedirect) {
+      const id = this.getAddonIdFromUrl(url);
+
+      // We shouldn't report built-in search engines.
+      if (!id.endsWith("@search.mozilla.org")) {
+        addonId = id;
+      }
+    }
+
+    if (!addonId) {
+      // No add-on ID means there is nothing we can report.
+      return;
+    }
+
+    // This is the (initial) URL before the redirect.
+    const from = await browser.addonsSearchExperiment.getPublicSuffix(url);
+
+    // This is the URL after the redirect.
+    const requestUrl =
+      redirectUrl ||
+      (await browser.addonsSearchExperiment.getRequestUrl(requestId));
+    const to = await browser.addonsSearchExperiment.getPublicSuffix(requestUrl);
+
+    if (from === to) {
+      // We do not report redirects to same public suffixes.
+      return;
+    }
```

Comment:
I'm not certain that we shouldn't log this.

Reply:
I don't think we're interested in extensions redirecting from `example.com` to `example.com` or `foo.example.com` (ETLD still being `example.com`) to `example.com` because we "stay" on the same domain.

cc @wagnerand

Reply:
At least for the first iteration, we want to log only when the domain is different. Non-domain changes in the url need more discussion and aligning before we can add logging for them.

Reply:
I feel like if some other extension is changing the request we should note that.  So it's not about a redirect from a.com to a.com, but a redirect from a.com to a.com caused by some random extension.  I understand that isn't apparent to users, but to me it does indicate an extension messing with a search request.